spring-integration-core/src/main/java/org/springframework/integration/handler/SuspendingMessageHandler.java

```diff
@@ -0,0 +1,132 @@
+
+	private final long timeout;
+
+	private final CorrelationStrategy correlationStrategy;
+
+	public SuspendingMessageHandler(long timeout) {
+		this(new HeaderAttributeCorrelationStrategy(IntegrationMessageHeaderAccessor.CORRELATION_ID), timeout);
+	}
+
+	public SuspendingMessageHandler(CorrelationStrategy correlationStrategy,
+			long timeout) {
+		this.correlationStrategy = correlationStrategy;
+		this.timeout = timeout;
+	}
+
+	@Override
+	protected Object handleRequestMessage(Message<?> requestMessage) {
+		Object key = SuspendingMessageHandler.this.correlationStrategy.getCorrelationKey(requestMessage);
+		if (key == null) {
+			throw new MessagingException(requestMessage, "Correlation Strategy returned null");
+		}
+		Thread existing = this.inProcess.putIfAbsent(key, Thread.currentThread());
+		if (existing != null) {
+			throw new MessagingException(requestMessage, "Correlation key ("
+					+ key + ") is already in use by " + existing.getName());
+		}
+		BlockingQueue<Message<?>> blockingQueue = createOrObtainQueue(key);
+		try {
+			Message<?> release = blockingQueue.poll(this.timeout, TimeUnit.MILLISECONDS);
+			if (release != null) {
+				this.suspensions.remove(key);
+				if (release.getPayload() instanceof Throwable) {
+					throw new MessagingException(requestMessage, "Releasing flow returned a throwable",
+							(Throwable) release.getPayload());
+				}
+				else {
+					return buildResult(requestMessage, release);
+				}
+			}
+			else {
+				synchronized(blockingQueue) {
+					if (!blockingQueue.offer(requestMessage)) {
+						// second chance
+						return buildResult(requestMessage, blockingQueue.poll());
+					}
+				}
+				return null;
+			}
+		}
+		catch (InterruptedException e) {
+			Thread.currentThread().interrupt();
+			throw new MessagingException(requestMessage, "Interrupted while waiting for release", e);
+		}
+		finally {
+			this.inProcess.remove(key);
+		}
+	}
+
+	private Object buildResult(Message<?> requestMessage, Message<?> release) {
+		Object[] result = new Object[2];
+		result[0] = requestMessage.getPayload();
+		result[1] = release;
+		return result;
+	}
+
+	private BlockingQueue<Message<?>> createOrObtainQueue(Object key) {
+		BlockingQueue<Message<?>> blockingQueue = new ArrayBlockingQueue<Message<?>>(1);
+		BlockingQueue<Message<?>> existing = this.suspensions.putIfAbsent(key, blockingQueue);
+		if (existing != null) {
+			blockingQueue = existing;
+		}
+		return blockingQueue;
+	}
+
+	public void release(Message<?> message) {
+		Object key = this.correlationStrategy.getCorrelationKey(message);
+		if (key == null) {
+			throw new MessagingException(message, "Correlation Strategy returned null");
+		}
+		BlockingQueue<Message<?>> blockingQueue = createOrObtainQueue(key);
+		synchronized(blockingQueue) {
+			if (!blockingQueue.offer(message)) {
+				this.logger.error("Release message arrived too late: " + message
+						+ " pending message: " + blockingQueue.poll());
+				this.suspensions.remove(key);
```

Comment:
OK. I see now your point regarding that `second chance`.
Enough smart, though...
But seems for me we can end up with `OOM`, if we allow to retain unreleased/cleared `suspensions`.
I mean that case, when the `releaseMessage` won't be sent at all.

Your thoughts ?